File: python/v9_symbolic.py
"""Deterministic allowed-values transition evaluator used by V9 scoring."""

from __future__ import annotations

from itertools import product
from typing import Any


VALUE_ORDER = ("inactive", "active")


def evaluate_allowed_transitions(
    schema: dict[str, Any],
    determinant_values: list[dict[str, Any]],
) -> dict[str, Any]:
    determinant_ids = [value["id"] for value in schema["transition_determinants"]]
    if len(set(determinant_ids)) != len(determinant_ids):
        raise ValueError("Transition schema contains duplicate determinant ids")
    values_by_id: dict[str, tuple[str, ...]] = {}
    for grounding in determinant_values:
        determinant_id = grounding["determinant_id"]
        if determinant_id not in determinant_ids:
            raise ValueError(f"Grounding contains unknown determinant {determinant_id}")
        if determinant_id in values_by_id:
            raise ValueError(f"Grounding repeats determinant {determinant_id}")
        supplied = grounding["allowed_values"]
        if any(value not in VALUE_ORDER for value in supplied):
            raise ValueError("Allowed values must be active or inactive")
        normalized = tuple(value for value in VALUE_ORDER if value in supplied)
        if not normalized:
            raise ValueError(f"Grounding has no allowed values for {determinant_id}")
        values_by_id[determinant_id] = normalized
    for determinant_id in determinant_ids:
        if determinant_id not in values_by_id:
            raise ValueError(f"Grounding omits determinant {determinant_id}")
    cases: dict[tuple[str, ...], str] = {}
    for transition_case in schema["transition_cases"]:
        values = tuple(transition_case["values"])
        if len(values) != len(determinant_ids):
            raise ValueError("Transition case arity differs from determinant schema")
        if values in cases:
            raise ValueError(f"Transition schema repeats assignment {values}")
        cases[values] = transition_case["transition_code"]
    assignments = list(product(*(values_by_id[value] for value in determinant_ids)))
    codes = []
    for assignment in assignments:
        if assignment not in cases:
            raise ValueError(f"Transition schema omits compatible assignment {assignment}")
        codes.append(cases[assignment])
    possible = sorted(set(codes))
    return {
        "compatible_assignments": len(assignments),
        "possible_transition_codes": possible,
        "identifiable": len(possible) == 1,
    }

File: src/v9-symbolic.ts
import type { V8StructuredInput } from './contracts';
import type {
  V9AllowedValues,
  V9BooleanValue,
  V9SymbolicInput,
  V9SymbolicResult,
} from './v9-contracts';

const VALUE_ORDER: V9BooleanValue[] = ['inactive', 'active'];

export function evaluateAllowedTransitions(input: V9SymbolicInput): V9SymbolicResult {
  const schema = input.action_dependency_schema;
  const determinantIds = schema.transition_determinants.map((value) => value.id);
  if (new Set(determinantIds).size !== determinantIds.length) {
    throw new Error('Transition schema contains duplicate determinant ids.');
  }
  const valuesById = new Map<string, V9BooleanValue[]>();
  for (const grounding of input.determinant_values) {
    if (!determinantIds.includes(grounding.determinant_id)) {
      throw new Error(`Grounding contains unknown determinant ${grounding.determinant_id}.`);
    }
    if (valuesById.has(grounding.determinant_id)) {
      throw new Error(`Grounding repeats determinant ${grounding.determinant_id}.`);
    }
    const values = normalizedValues(grounding.allowed_values);
    if (values.length === 0) {
      throw new Error(`Grounding has no allowed values for ${grounding.determinant_id}.`);
    }
    valuesById.set(grounding.determinant_id, values);
  }
  for (const id of determinantIds) {
    if (!valuesById.has(id)) throw new Error(`Grounding omits determinant ${id}.`);
  }

  const caseByAssignment = new Map<string, string>();
  for (const transitionCase of schema.transition_cases) {
    if (transitionCase.values.length !== determinantIds.length) {
      throw new Error('Transition case arity differs from its determinant schema.');
    }
    const key = assignmentKey(transitionCase.values);
    if (caseByAssignment.has(key)) throw new Error(`Transition schema repeats assignment ${key}.`);
    caseByAssignment.set(key, transitionCase.transition_code);
  }

  const compatible = cartesianProduct(
    determinantIds.map((id) => required(valuesById, id)),
  );
  const codes = compatible.map((assignment) => {
    const key = assignmentKey(assignment);
    const code = caseByAssignment.get(key);
    if (code === undefined) throw new Error(`Transition schema omits compatible assignment ${key}.`);
    return code;
  });
  const possible = [...new Set(codes)].sort();
  return {
    compatible_assignments: compatible.length,
    possible_transition_codes: possible,
    identifiable: possible.length === 1,
  };
}

export function allowedValuesFromV8Input(input: V8StructuredInput): V9AllowedValues[] {
  const evidenceById = new Map(input.evidence_ledger.map((fact) => [fact.id, fact]));
  return input.action_dependency_schema.transition_determinants.map((determinant) => {
    const evidence = evidenceById.get(determinant.id);
    if (!evidence) throw new Error(`V8 evidence omits determinant ${determinant.id}.`);
    if (evidence.evidence_state === 'unresolved') {
      return { determinant_id: determinant.id, allowed_values: [...VALUE_ORDER] };
    }
    if (evidence.value !== 'active' && evidence.value !== 'inactive') {
      throw new Error(`Confirmed V8 determinant ${determinant.id} has no Boolean value.`);
    }
    return { determinant_id: determinant.id, allowed_values: [evidence.value] };
  });
}

function normalizedValues(values: V9BooleanValue[]): V9BooleanValue[] {
  if (values.some((value) => value !== 'active' && value !== 'inactive')) {
    throw new Error('Allowed values must be active or inactive.');
  }
  return VALUE_ORDER.filter((value) => values.includes(value));
}

function cartesianProduct(values: V9BooleanValue[][]): V9BooleanValue[][] {
  return values.reduce<V9BooleanValue[][]>(
    (prefixes, next) => prefixes.flatMap((prefix) => next.map((value) => [...prefix, value])),
    [[]],
  );
}

function assignmentKey(values: readonly V9BooleanValue[]): string {
  return values.join('|');
}

function required<K, V>(values: Map<K, V>, key: K): V {
  const value = values.get(key);
  if (value === undefined) throw new Error(`Missing required map value ${String(key)}.`);
  return value;
}
